compiler/src/dmd/expressionsem.d

```diff
@@ -13416,8 +13429,8 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             __equals = __equals.trySemantic(sc); // for better error message
             if (!__equals)
             {
-                error(exp.loc, "incompatible types for array comparison: `%s` and `%s`",
-                          exp.e1.type.toChars(), exp.e2.type.toChars());
+                if(sc.func)
+                    error(exp.loc, "can't infer return type in function `%s`", sc.func.toChars());
                 __equals = ErrorExp.get();
```

Comment:
either `sc.func` should always be true or you want
```d
if (!__equals)
{
    if (sc.func)
        error(exp.loc, "can't infer return type in function `%s`", sc.func.toChars());
    else
        error(exp.loc, "incompatible types for array comparison: `%s` and `%s`",
                  exp.e1.type.toChars(), exp.e2.type.toChars());
    __equals = ErrorExp.get();
}
```
here. A path that returns an `ErrorExp` should always issue an error.
Also, space between `if (` for `if(sc.func)`.